### ISSUE TITLE
Investigate playability issues with PaybackTime 2 (1997)

### ISSUE THREAD

**Issue by kcgen** [dosbox-staging/dosbox-staging#2920]
### Context

Currently the eXoDOS project runs this game using [the DOSBox Daum fork](https://github.com/ykhwong/dosbox-svn-daum), which is understood to have specific emulation fixes not currently in upstream DOSBox (0.74.x and SVN) or active forks such as X, Pure, and Staging.

DOSBox Daum had 51 releases that included DOSBox SVN's commits ~r3600 through ~r3900, [spanning 2010 through 2015](http://ykhwong.x-y.net/).

### The goal of this ticket is to answer the following

Is Daum _still_ needed for this game?
- [ ] Does DOSBox SVN ([download latest](http://www.emucr.com/2023/09/dosbox-svn-r4483.html)) play it OK?
- [ ] Does DOSBox X ([download latest](https://dosbox-x.com/)) play it OK?
- [ ] Does DOSBox Staging ([download latest](https://dosbox-staging.github.io/downloads/development-builds/)) play it OK?

If Daum plays it better:
- [ ] What aspect is better?
- [ ] Which Daum release was the **_first_** to fix it?
    Use these builds [here](http://ykhwong.x-y.net/subpages/old.php) to narrow it down.

### How you can help

 1. Download the above lastest DOSBox SVN, X, and Staging builds.
 1. Ensure you have [eXoDOS v5](https://www.retro-exo.com/exodos.html) (or eventually v6)
 1. Find this game inside the eXoDOS installed folders (look for: `PBTime2/`)
 1. Run the game using the DOSBox SVN, X, and Staging builds.
    _If it plays fine (everything looks, sounds, and handles OK) then simply indicate as such. Your task is done!_

If SVN, X, and Staging all have problems, then try it with Daum:
 1. Download the [first and last Daum releases](http://ykhwong.x-y.net/subpages/old.php).
    Note: _Daum had the most releases for Microsoft Windows, so having a Windows system (or VM) will help the most to answer the last question above._
 1. Does the latest Daum release play it correctly? If so: summarize/describe how it's better than DOSBox SVN, X, and Staging (screenshots, video, audio records all help too).
 1. Find the earliest (or oldest) Daum release that fixed it.
    1. Run the Daum release half way between the known good release (where the game playing correctly) and bad release (where the game doesn't playing correctly) until you've narrowed it down to adjacent releases.
    1. It will take _at-most_ seven runs of different Daum versions to narrow this down (using this middle-selecting technique); so it's not as daunting as it seems! :-)

---

### Follow up by code contributors

If Daum plays it better, your answers (especially the last one that narrows down the fix to a specific release) are critical to helping isolate the chunk of code fixed the given bug.

Pull requests should credit:
- @ykhwong as the original code author,
- the test team member who confirmed and isolated the Daum release,
- yourself (PR author) as integrating the fix

The PR should include the `imported-from:` URL pointing to DOSBox Daum. For an example, see commit a35c7c889.

**Comment by ebenbruyns** [comment]
I have done limited testing on this one, so far on staging I get a hum when the music stops after placing your players, this is not present in daum on linux. I just wanted to put this here so it's tracked somewhere. I have more testing to do.

**Comment by johnnovak** [comment]
> I have done limited testing on this one, so far on staging I get a hum when the music stops after placing your players, this is not present in daum on linux

This is with OPL sound, right? I noticed that OPL music/sound doesn't get fully muted when there is supposed to be complete silence in a few games. It's tricky no notice because usually you can only hear it on headphones. It happens in too many games for this to be a game specific bug; I suspect we introduced a regression a while ago. Maybe it has something to do with the auto channel muter stuff. I could notice this in Beneath A Steel Sky with OPL sound and a few other games I forgot.

**Comment by ebenbruyns** [comment]
> > I have done limited testing on this one, so far on staging I get a hum when the music stops after placing your players, this is not present in daum on linux
> 
> This is with OPL sound, right? I noticed that OPL music/sound doesn't get fully muted when there is supposed to be complete silence in a few games. It's tricky to notice because usually you can only hear it on headphones. It happens in too many games for this to be a game specific bug; I suspect we introduced a regression a while ago. Maybe it has something to do with the auto channel muter stuff. I could notice this in Beneath A Steel Sky with OPL sound and a few other games I forgot.

I think so. I'm going to try apply that opl change I found in duam properly and see what happens, but that wont be a today job...

**Comment by johnnovak** [comment]
> > > I have done limited testing on this one, so far on staging I get a hum when the music stops after placing your players, this is not present in daum on linux
> > 
> > 
> > This is with OPL sound, right? I noticed that OPL music/sound doesn't get fully muted when there is supposed to be complete silence in a few games. It's tricky to notice because usually you can only hear it on headphones. It happens in too many games for this to be a game specific bug; I suspect we introduced a regression a while ago. Maybe it has something to do with the auto channel muter stuff. I could notice this in Beneath A Steel Sky with OPL sound and a few other games I forgot.
> 
> I think so. I'm going to try apply that opl change I found in duam properly and see what happens, but that wont be a today job...

Or compare it to SVN and X; they might not have the bug (if it's a bug).

**Comment by Grounded0** [comment]
@johnnovak is this something introduced after 0.80.1? Hard to keep up when versions aren't reported.

**Comment by Grounded0** [comment]
@johnnovak Main reason why its difficult to detect without high-end headphones is the fact that its usually extremely low frequency hanging note that produces discomfort to me that i only notice when i terminate DOSBox Staging. Its out from frequency response band of almost anything that costs less than 200 USD so i got screwed over again by buying expensive shit. 😁

**Comment by johnnovak** [comment]
> @johnnovak Main reason why its difficult to detect without high-end headphones is the fact that its usually extremely low frequency hanging note that produces discomfort to me that i only notice when i terminate DOSBox Staging. Its out from frequency response band of almost anything that costs less than 200 USD so i got screwed over again by buying expensive shit. 😁

Yeah @Grounded0, I can clearly notice it in some games on my studio headphones too. Beneath the Steel Sky is one of them.

Well, in this game it's not just a "hum" after starting the game; the notes are literally stuck forever and still playing, it's very loud. We might be conflating different issues here, so I suggest to raise separate "hanging note on OPL" tickets for the sound issues. The hanging note bug in this game is present in SVN too. I will try it on real hardware at some point.

> Is this something introduced after 0.80.1?

The problem with BASS is present in 0.80.1 and in current main. It's not present in SVN.

The hanging note in this game is present both on SVN, 0.80.1 and current main.

That's why I recommend separate, distinct tickets for these issues. Can you raise one for this game @ebenbruyns or @Grounded0 ?

**Comment by Grounded0** [comment]
I'll first chew down this list down a bit at least all AAA titles.

**Comment by kcgen** [comment]
> The hanging note in this game is present both on SVN, 0.80.1 and current main.

Nice.. if the hanging note doesn't exist in Circa-2015 Daum, then it will bisect into two places:
 1. A regression introduced in the range of SVN commits post-2015 Daum and present day, or
 2. A fix in Daum, which will be revealed in @ebenbruyns's coarse reconstruction of the Daum-only additions to SVN.

**Comment by johnnovak** [comment]
Yeah, another avenue is to test the problematic games in PCem.

**Comment by ebenbruyns** [comment]
I'm not so sure that daum solves the hanging note, I think I just hit enter at the RIGHT time and it wasn't playing at the time

**Comment by johnnovak** [comment]
> I'm not so sure that daum solves the hanging note, I think I just hit enter at the RIGHT time and it wasn't playing at the time

Cheers, that's some good info there.

**Comment by kcgen** [comment]
Daum:

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/c6a96c8a-0f48-4c61-938b-4872c376c584

Staging `main`, all defaults:


https://github.com/dosbox-staging/dosbox-staging/assets/1557255/d33802ba-189c-43f4-9b47-834f1b19bd2e


https://github.com/dosbox-staging/dosbox-staging/assets/1557255/5b8cbc39-a79a-4a53-976f-5d1f693a334c

**Comment by ebenbruyns** [comment]
when you start your turn the game says "Go" it it chopped off on staging but daum gets that part right.

**Comment by ebenbruyns** [comment]
Another thing is with just one human player I get the hanging note, with 2 human players the music keeps going. But that seems to be across the board so far

**Comment by ebenbruyns** [comment]
the game keeps hanging for me just trying to play a single turn. Might be my settings, anybody else get the same behaviour? This happens on staging, daum linux and daum on wine

**Comment by ebenbruyns** [comment]
doing a bit of digging I found that the author has released the source code for this game: https://github.com/nnevatie/paybacktime2

migth come in handy

**Comment by johnnovak** [comment]
> @johnnovak Is this something introduced after 0.80.1? Hard to keep up when versions aren't reported. I think i've noticed this on my Sennheiser IE400 Pro with external DAC. Only after terminating DOSBox Staging does the hanging note terminate. This should become its own ticket.

@Grounded0 I'm going to create a ticket for the "hanging note" issue but I need as many repro cases as possible. I have a few that I will test on real hardware shortly (e.g. Beneath a Steel Sky).

Can you list a few repro cases for the OPL hanging note problem? Just the title and a few words how it manifests is enough, then I'll do the investigation and turn it into a ticket. Thanks!

**Comment by Grounded0** [comment]
> > @johnnovak Is this something introduced after 0.80.1? Hard to keep up when versions aren't reported. I think i've noticed this on my Sennheiser IE400 Pro with external DAC. Only after terminating DOSBox Staging does the hanging note terminate. This should become its own ticket.
> 
> @Grounded0 I'm going to create a ticket for the "hanging note" issue but I need as many repro cases as possible. I have a few that I will test on real hardware shortly (e.g. Beneath a Steel Sky).
> 
> Can you list a few repro cases for the OPL hanging note problem? Just the title and a few words how it manifests is enough, then I'll do the investigation and turn it into a ticket. Thanks!

I currently don't have the capacity for this due to serious mental illness but anyone else please go ahead with this.

**Comment by johnnovak** [comment]
> > @Grounded0 I'm going to create a ticket for the "hanging note" issue but I need as many repro cases as possible. I have a few that I will test on real hardware shortly (e.g. Beneath a Steel Sky).
> > Can you list a few repro cases for the OPL hanging note problem? Just the title and a few words how it manifests is enough, then I'll do the investigation and turn it into a ticket. Thanks!
> 
> I currently don't have the capacity for this due to [serious mental illness](https://en.wikipedia.org/wiki/Serious_mental_illness) but anyone else please go ahead with this.

Hope you'll better soon, man.

**Comment by nnevatie** [comment]
> Plus this all assumes that the game is not broken and does _not_ have the hanging note bug on real hardware.

Hi, original author here. IIRC real hardware did not reproduce the hanging note bug. The game was mostly tested on SB1, SB2 and SB16 soundcards, with 386SX through 486 for CPUs.

Let me know if I can be of any help with this issue!

**Comment by nnevatie** [comment]
A further note regarding the game's music files. The files are named as `<filename>.SVM` - the files can be opened by Reality Adlib Tracker, which coincidentally is still maintained and can be downloaded from here: https://www.3eality.com/productions/reality-adlib-tracker

**Comment by MasterO2** [comment]
https://audacious-media-player.org can also open tracker files.

**Comment by johnnovak** [comment]
@nnevatie Appreciate the detailed repro steps, but the current SB emulation code works fine with the vast majority of the DOS gaming catalogue.

Changing the SB DSP code fundamentally is always risky as Creative's own documentation is very spotty, so most of the real-world behaviour had to be reverse engineered by the original DOSBox team.

Given we're in an unique postion with this game that you, the author, is contributing to this issue ticket, it would be a better and safer solution if you adjusted your SB code slightly so it works correctly in Staging (just like tha majority of DOS games), then released a new version of your game.

**Comment by Grounded0** [comment]
Game repo:

https://github.com/nnevatie/paybacktime2

**Comment by Python-Exoproject** [comment]
I was testing this game to see if it was viable to move from Daum to Staging. I got some screenshots of the crashes I experienced for Staging 81.1 while doing so which I post below in case its ever useful. I can also confirm that all issues went away when Music was disabled and that the issue exists for all dosbox forks with minor variations.

Random crash from just moving around the screen, also created a bunch of garbage files in the game folder:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/67406191/f219609e-7923-48af-89d1-69841416ed08)


Error given by game when try to shoot with any character and a paused mem /c to see how much memory was available:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/67406191/544d231a-b9d0-4d75-8403-d101969c260a)


Also just got complete freeze occasionally

**Comment by nnevatie** [comment]
Hi. I'm back from a time machine which I used to fix the original game. It took a while to find all the necessary toolchains and to decipher the 25+ years old code.

Could you try the latest binaries found in the game's repository after this commit? https://github.com/nnevatie/paybacktime2/commit/c46bc3aa03ea956c04971d3b7bc193bec67b7d51

Thanks for the thorough investigation of the original game's issues.

**Comment by Python-Exoproject** [comment]
Confirmed working using copy found at https://rflxn.com/paybacktime2/ in preliminary check. Well done @nnevatie and thank you for spending the time to get your game working again.

This can be closed @johnnovak

**Comment by Grounded0** [comment]
Closing on behalf of @johnnovak.

I remember seeing this game on Mikrobitti magazine in the 1990s @nnevatie. Glad it is now fixed for preservation purposes as there's not too many Finnish DOS games out there.

**Comment by johnnovak** [comment]
Great stuff @nnevatie! It's very cool you took the time and effort to go back in time and fix the code so it works in Staging too! 🥳
